False positive `unused-private-member` when using `__new__`

This may relate to #4657

### Steps to reproduce
Given a file `sample2.py`:

```python
class UnusedPrivateMember:
    """unused-private-member (W0238)"""

    def __new__(cls, func, *args):
        obj = super(UnusedPrivateMember, cls).__new__(cls)
        obj.func = func
        obj.__args = args
        return obj

    def exec(self):
        return self.func(*self.__args)
```

<!--
Given a file `a.py`:
```python
# (Disable message unrelated to the bug)
# pylint: disable=missing-docstring,...

```

Given multiple files:
```
__init__.py
a.py
module\
    __init__.py
    b.py
    c.py
```
-->

### Current behavior

Result of `pylint sample2.py`:
```
************* Module sample2
sample2.py:7:8: W0238: Unused private member `UnusedPrivateMember.__args` (unused-private-member)
```


### Expected behavior

I would expect no unused-private-member warning

### pylint --version output

Result of `pylint --version` output:

```
pylint 3.0.0-a4
astroid 2.6.2
Python 3.7.3 (default, Jan 22 2021, 20:04:44)
[GCC 8.3.0]
```

<!--
Additional dependencies:
```
pandas==0.23.2
marshmallow==3.10.0
...
```
-->

--- a/ChangeLog
+++ b/ChangeLog
@@ -67,7 +67,6 @@ Release date: TBA
   Closes #4668
 
 
-
 What's New in Pylint 2.9.3?
 ===========================
 Release date: 2021-07-01

--- a/ChangeLog
+++ b/ChangeLog
@@ -62,6 +62,11 @@ Release date: TBA
 
   Closes #4368
 
+* Fix false-positive of ``unused-private-member`` when using ``__new__`` in a class
+
+  Closes #4668
+
+
 
 What's New in Pylint 2.9.3?
 ===========================

--- a/pylint/checkers/classes.py
+++ b/pylint/checkers/classes.py
@@ -964,15 +964,14 @@ a metaclass class method.",
             # Logic for checking false positive when using __new__,
             # Get the returned object names of the __new__ magic function
             # Then check if the attribute was consumed in other instance methods
-            acceptable_obj_names = ["self"]
-            if (
-                isinstance(assign_attr.scope(), astroid.FunctionDef)
-                and assign_attr.scope().name == "__new__"
-            ):
+            acceptable_obj_names: List[str] = ["self"]
+            scope = assign_attr.scope()
+            if isinstance(scope, astroid.FunctionDef) and scope.name == "__new__":
                 acceptable_obj_names.extend(
                     [
-                        n.value.name
-                        for n in assign_attr.scope().nodes_of_class(astroid.Return)
+                        return_node.value.name
+                        for return_node in scope.nodes_of_class(astroid.Return)
+                        if isinstance(return_node.value, astroid.Name)
                     ]
                 )
 

--- a/pylint/checkers/classes.py
+++ b/pylint/checkers/classes.py
@@ -960,6 +960,22 @@ a metaclass class method.",
         for assign_attr in node.nodes_of_class(astroid.AssignAttr):
             if not is_attr_private(assign_attr.attrname):
                 continue
+
+            # Logic for checking false positive when using __new__,
+            # Get the returned object names of the __new__ magic function
+            # Then check if the attribute was consumed in other instance methods
+            acceptable_obj_names = ["self"]
+            if (
+                isinstance(assign_attr.scope(), astroid.FunctionDef)
+                and assign_attr.scope().name == "__new__"
+            ):
+                acceptable_obj_names.extend(
+                    [
+                        n.value.name
+                        for n in assign_attr.scope().nodes_of_class(astroid.Return)
+                    ]
+                )
+
             for attribute in node.nodes_of_class(astroid.Attribute):
                 if attribute.attrname == assign_attr.attrname and (
                     (
@@ -968,7 +984,11 @@ a metaclass class method.",
                         and attribute.expr.name in ["cls", "self"]
                     )
                     # If assigned to self.attrib, can only be accessed by self
-                    or (assign_attr.expr.name == attribute.expr.name == "self")
+                    # Or if __new__ was used, the returned object names are acceptable
+                    or (
+                        assign_attr.expr.name in acceptable_obj_names
+                        and attribute.expr.name == "self"
+                    )
                 ):
                     break
             else:

--- a/tests/functional/u/unused/unused_private_member.py
+++ b/tests/functional/u/unused/unused_private_member.py
@@ -135,3 +135,26 @@ class FalsePositive4657:
     def attr_c(self):
         """Get c."""
         return cls.__attr_c  # [undefined-variable]
+
+# Test cases for false-positive reported in #4668
+# https://github.com/PyCQA/pylint/issues/4668
+
+class FalsePositive4668:
+    # pylint: disable=protected-access, no-member
+
+    def __new__(cls, func, *args):
+        if args:
+            true_obj = super(FalsePositive4668, cls).__new__(cls)
+            true_obj.func = func
+            true_obj.__args = args  # Do not emit message here
+            return true_obj
+
+        false_obj = super(FalsePositive4668, cls).__new__(cls)
+        false_obj.func = func
+        false_obj.__args = args  # Do not emit message here
+        false_obj.__secret_bool = False
+        return false_obj
+
+    def exec(self):
+        print(self.__secret_bool)
+        return self.func(*self.__args)

--- a/tests/functional/u/unused/unused_private_member.py
+++ b/tests/functional/u/unused/unused_private_member.py
@@ -136,11 +136,11 @@ class FalsePositive4657:
         """Get c."""
         return cls.__attr_c  # [undefined-variable]
 
+
 # Test cases for false-positive reported in #4668
 # https://github.com/PyCQA/pylint/issues/4668
-
 class FalsePositive4668:
-    # pylint: disable=protected-access, no-member
+    # pylint: disable=protected-access, no-member, unreachable
 
     def __new__(cls, func, *args):
         if args:
@@ -154,6 +154,8 @@ class FalsePositive4668:
         false_obj.__args = args  # Do not emit message here
         false_obj.__secret_bool = False
         return false_obj
+        # unreachable but non-Name return value
+        return 3+4
 
     def exec(self):
         print(self.__secret_bool)

